File: vaadin-testbench-it/src/main/java/com/vaadin/testbench/tests/ui/element/PolymerTemplateView.java
package com.vaadin.testbench.tests.ui.element;

import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.NativeButton;
import com.vaadin.flow.component.polymertemplate.PolymerTemplate;
import com.vaadin.flow.router.Route;
import com.vaadin.flow.templatemodel.TemplateModel;

@Route(PolymerTemplateView.ROUTE)
@Tag("polymer-template-view")
@JsModule("./polymer-template-view.js")
public class PolymerTemplateView extends PolymerTemplate<TemplateModel> {

    public static final String ROUTE = "PolymerTemplateView";

    public PolymerTemplateView() {
        for (int i = 1; i < 6; i++) {
            NativeButton button = new NativeButton("Button " + i);
            button.getElement().setAttribute("id", "light-button-" + i);
            button.addClassName("button");
            button.addClassName("button-" + i);
            getElement().appendChild(new Div(button).getElement());
        }

        NativeButton slottedButton = new NativeButton(
                "Special Button (in Light DOM)");
        slottedButton.getElement().setAttribute("slot", "special-slot");
        slottedButton.getElement().setAttribute("id", "special-button");
        slottedButton.addClassName("button");
        slottedButton.addClassName("button-special-slot");
        getElement().appendChild(slottedButton.getElement());
    }
}

File: vaadin-testbench-it/frontend/polymer-template-view.js
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';

class PolymerTemplateView extends PolymerElement {
    static get is() {
	return 'polymer-template-view';
    }
    static get template() {
	return html`
	    <div>
			<button class="button button-shadow-1" id="shadow-button-1">Shadow Button 1</button>
	    </div>
	    <slot>XXX</slot>
	    <div>
			<button class="button button-shadow-2" id="shadow-button-2">Shadow Button 2</button>
	    </div>
	    <slot name="special-slot"></slot>
	    <div>
			<button class="button button-shadow-special" id="special-button">Special Button (in Shadow DOM)</button>
	    </div>
	    <button class="button" id="foo'*+bar'">Button with special id</button>`;
    }
}

window.customElements.define(PolymerTemplateView.is, PolymerTemplateView);

